Allow changing the bot type of a bot user in the "manage bot" modal.

Varun-Kolanu opened the pull request:
<!-- Describe your pull request here.-->

Fixes #31389

Previously, there was no option for a user to edit their bot type.
This was leading to a long workaround of deactivating the bot -> creating a new bot with the required bot type.

This PR attempts to introduce the feature of changing the bot type through Manage bot modal.

## Commits Explanations:
1. Adds event handler for editing an embedded bot service.
2. Previously, the initial fetch state of events wasn't returning the services of an embedded bot, if there are no config options available. This commit fixes it (since we need service_name of the bot in frontend to let users change it)
3. Backend changes related to bot_type change and changing embedded bot service name.
4. The main commit (6th one) needed importing `get_allowed_bot_types` in the `user_profile.ts` and it was causing a circular dependency error. Hence I moved that function (and related functions) in the 4th commit from `settings_bots.ts` into `user_profile.ts` to avoid the error.
5. Changed the types of `bot_types` constant for maintaining uniformity
6. Main commit, which supports editing the bot type, and also supports the frontend changes for editing embedded bot service name. 

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**


https://github.com/user-attachments/assets/a0c0fea4-e7b5-4c0c-ad26-55d3c1bbd75d



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Varun-Kolanu committed:
integrations: Move `get_allowed_bot_types` function.

Move the `get_allowed_bot_types` and required functions
into `user_profile.ts` to avoid circular dependencies.

Preparatory commit for #31389.

Varun-Kolanu committed:
integrations: Add backend support for changing bot type.

This commit adds backend support to update a bot's type
by accepting the `bot_type` parameter through the
`/json/bots/{bot_id}` endpoint. It also adds support
for accepting the `embedded_bot_service_name` parameter
through the same endpoint.

Fixes part of #31389.

Co-authored-by: Pieter Cardillo Kwok <pieterceka123@gmail.com>

Varun-Kolanu commented on web/src/bot_data.ts:
```diff
@@ -43,8 +52,12 @@ export function update(bot_id: number, bot_update: ServerUpdateBotData): void {
+            Object.assign(service, services_update[0]);
+        } else {
+            services.set(bot_id, services_update);
+        }
```
This particular `else` code block is helpful when bot_type is being changed to OUTGOING_WEBHOOK when initially there was no service related to that bot.

Varun-Kolanu commented on web/src/bot_data.ts:
```diff
@@ -68,6 +81,45 @@ export function get_all_bots_owned_by_user(user_id: number): Bot[] {
     return ret;
 }
 
+export function can_create_new_bots(): boolean {
```
This file seemed to be apt for moving these functions to avoid circular dependency. Any other suitable file suggestion would also be appreciated.

Varun-Kolanu commented on web/src/user_profile.ts:
```diff
@@ -89,8 +90,8 @@ let bot_owner_dropdown_widget: DropdownWidget | undefined;
 const INCOMING_WEBHOOK_BOT_TYPE = 2;
-const OUTGOING_WEBHOOK_BOT_TYPE = "3";
-const EMBEDDED_BOT_TYPE = "4";
+const OUTGOING_WEBHOOK_BOT_TYPE = 3;
```
These can be in `int` type since we are now dealing the `bot_type` in `int` type only.

timabbott replied:
Can you make the refactor that makes this valid its own prep commit? I don't remember why it was a string and the ideal way to communicate your understanding of that is to have a commit just changing these to int with a commit message explaining why that's correct.

Varun-Kolanu commented on zerver/actions/users.py:
```diff
@@ -780,15 +782,27 @@ def do_change_can_change_user_emails(user_profile: UserProfile, value: bool) ->
-    service.interface = service_interface
-    service.save()
+    services = get_bot_services(bot_profile.id)
+    if len(services) == 0:
```
This block is helpful when there is no service initially present, but the bot type is being changed to `OUTGOING_WEBHOOK_BOT`

Varun-Kolanu commented:
@zulipbot add "buddy review"

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,11 @@ format used by the Zulip server that they are interacting with.
 
+**Feature level 393**
+
+* `PATCH /bots/{bot_user_id}`: Added support to change a bot's type
```
Please use the `tools/create-api-changelog` to avoid merge conflicts here.

timabbott commented on zerver/tests/test_bots.py:
```diff
@@ -1524,6 +1524,47 @@ def test_patch_bot_role(self) -> None:
+
+        req = dict(bot_type=UserProfile.EMBEDDED_BOT)
+        result = self.client_patch(f"/json/bots/{bot_profile.id}", req)
+        self.assert_json_success(result)
```
Whenever you have an `assert_json_success`, you need to also be checking that the change actually took effect properly. A good technique would be to `GET` the bot user and inspect the properties of what was returned.

timabbott commented on zerver/actions/users.py:
```diff
@@ -826,6 +842,41 @@ def do_update_bot_config_data(bot_profile: UserProfile, config_data: dict[str, s
+    previous_bot_type = bot_profile.bot_type
+
+    bot_profile.bot_type = bot_type
+    bot_profile.save(update_fields=["bot_type"])
```
Do we have any caches of bots that may need special flushing logic? I'd expect no, but `git grep cache_with_key` should help audit.

timabbott commented on web/src/settings_bots.ts:
```diff
@@ -45,11 +43,6 @@ type BotInfo = {
-type BotType = {
-    type_id: number;
-    name: string;
-};
```
Would `bot_data.ts` be more appropriate for where to move this?

timabbott commented on web/src/user_profile.ts:
```diff
@@ -836,12 +827,44 @@ export function show_edit_bot_info_modal(user_id: number, $container: JQuery): v
+    let bot_type: number = bot.bot_type;
+
+    render_bot_edit_form_based_on_bot_type();
+    function render_bot_edit_form_based_on_bot_type(): void {
```
I'd put this in a comment at the start of the function :)

Varun-Kolanu committed:
integrations: Change the types of bot_type constants to integer.

For consistency in bot_type constant types, this commit changes
the types of `OUTGOING_WEBHOOK_BOT_TYPE` and
`EMBEDDED_BOT_TYPE` from `string` to `int`.
This allows us to directly use `bot_type` as an integer in
comparisons with these constants.

Varun-Kolanu commented on zerver/actions/users.py:
```diff
@@ -826,6 +842,41 @@ def do_update_bot_config_data(bot_profile: UserProfile, config_data: dict[str, s
+        bot_profile.realm,
+        dict(type="realm_bot", op="update", bot=payload),
+        bot_owner_user_ids(bot_profile),
+    )
```
Added in the new update

PieterCK commented:
I can review this tomorrow!

PieterCK reviewed (commented):
@Varun-Kolanu I think this is great! I did some manual testings, and it works pretty much as expected. I left some comments on some code details. Other than that, I think:
- We should add some tests that check whether the bot functions as expected when it has changed bot type. Probably in `test_service_bot_system.py`.
- It seems like we can move some of the preparatory changes in the last commit that updates the frontend infrastructure to its own commit.
- In the last commit, the commit description and summary feels like it's more of a PR description than description for those changes. I would suggest something along the lines of: `user_profile: Add UI to change bot users bot_type.` and that commit could just be about adding the "Bot type" input field and the logic related to that in `user_profile.ts`.

Let me know if any of my comments are unclear!

timabbott commented:
This looks like it won't be finished for 11.0. But let's aim to get it finished soon; it shouldn't be very complex.

Varun-Kolanu commented:
Hi @PieterCK!
Can you have a quick pass on the PR once and LMK if it can be passed to the next phase of review? I've updated it according to the reviews.

PieterCK commented on zerver/views/users.py:
```diff
@@ -497,6 +501,37 @@ def patch_bot_backend(
+                # Embedded bots `service_name` must be one of EMBEDDED_BOTS
+                # TODO: Update this logic once we want to add support for
+                # updating embedded bots.
+                service_name = EMBEDDED_BOTS[0].name
```
Hmm, this seems to not include the validation and logic for an embedded bots Service. We probably want to just raise a `JsonableError` here if we're not planning to support updating the bot type to Embedded bot for now? Something like:
```python
raise JsonableError(_("Unable to change bot type to embedded bot."))
```

PieterCK commented on zerver/tests/test_service_bot_system.py:
```diff
@@ -629,3 +630,48 @@ def test_flag_messages_service_bots_has_processed(self) -> None:
+            service_interface=Service.GENERIC,
+        )
+
+        content = "@**FooBot** foo bar!!!"
```
This looks like an error, I think we want to mention the new bot you just created in this test and not `FooBot`.
`content = "@**GenBot** hi hi`

PieterCK commented:
@Varun-Kolanu Thanks for making those changes! I've left some more comments, but this looks like it has addressed the majority of comments left by Tim and me. I think you can pass this to the next reviewers once you've addressed the new comments from me.

Manual tests I did:
- changing bot type from generic bot to outgoing
- changing the same bot back to generic and making sure it doesn't keep the previous bot types function
- changing the same bot back again to the outgoing webhook bot to see it still has the correct `Service` data
- changing the bot type to incoming webhook bot.

None of the manual tests above went wrong when I tested.

An unresolved discussion that I'd like to highlight:
- Support changing to embedded bot? https://github.com/zulip/zulip/pull/34929#discussion_r2247726038.

Varun-Kolanu committed:
bots: Add event handler to update embedded bot service.

This is a preparatory commit for adding support
to editing embedded bot services. The handler
accepts a new name for the service.

Varun-Kolanu committed:
bots: Fix mismatched initial state of embedded bot service.

This commit ensures that the backend returns the services
of an embedded bot even when no config_data is present.

Varun-Kolanu committed:
integrations: Add UI support for editing bot type and service name.

This commit updates the Manage Bots modal to allow editing
a bot's `bot_type`. It also adds support for editing the
`service_name` of an embedded bot.

Fixes #31389.

Varun-Kolanu commented:
It's ready for review now. I addressed the following things from the previous review:
1. Added support for changing bot into embedded bot, by adding event handlers to take care of editing an embedded bot service by changing its service_name.
2. Updated the test in `test_services_bot_system` according to the review.

I have updated the recording of the video in the PR description too.

Varun-Kolanu commented:
@zulipbot add "mentor review"
